c/serverStatusService.c

```diff
@@ -0,0 +1,262 @@
+}
+
+void respondWithServerRoutes(HttpResponse *response){
+  jsonPrinter *out = respondWithJsonPrinter(response);
+  setResponseStatus(response, 200, "OK");
+  setDefaultJSONRESTHeaders(response);
+  writeHeader(response);
+  jsonStart(out);
+  jsonStartArray(out, "links");
+  jsonStartObject(out, NULL);
+  jsonAddString(out, "href", "/server/agent/config");
+  jsonAddString(out, "rel", "config");
+  jsonAddString(out, "type", "GET");
+  jsonEndObject(out);
+  jsonStartObject(out, NULL);
+  jsonAddString(out, "href", "/server/agent/log");
+  jsonAddString(out, "rel", "log");
+  jsonAddString(out, "type", "GET");
+  jsonEndObject(out);
+  jsonStartObject(out, NULL);
+  jsonAddString(out, "href", "/server/agent/logLevels");
+  jsonAddString(out, "rel", "logLevels");
+  jsonAddString(out, "type", "GET");
+  jsonEndObject(out);
+  jsonStartObject(out, NULL);
+  jsonAddString(out, "href", "/server/agent/environment");
+  jsonAddString(out, "rel", "environment");
+  jsonAddString(out, "type", "GET");
+  jsonEndObject(out);
+  jsonEndArray(out);
+  jsonEnd(out);
+  finishResponse(response);
+}
+
+void respondWithLogLevels(HttpResponse *response, ServerAgentContext *context){
+  jsonPrinter *out = respondWithJsonPrinter(response);
+  JsonObject *logLevels = jsonObjectGetObject(context->serverConfig, "logLevels");
+  setResponseStatus(response, 200, "OK");
+  setDefaultJSONRESTHeaders(response);
+  writeHeader(response);
+  jsonStart(out);
+  if(logLevels){
+    jsonPrintObject(out, logLevels);
+  }
+  jsonEnd(out);
+  finishResponse(response);
+}
+
+void respondWithServerEnvironment(HttpResponse *response, ServerAgentContext *context){
+  /*Information about parameters for smf_unc: https://www.ibm.com/support/knowledgecenter/SSLTBW_2.1.0/com.ibm.zos.v2r1.erbb700/smfp.htm#smfp*/
+  struct utsname unameRet;
+  char smfVarBuffer[64];
+  char hostnameBuffer[256];
+  char* buffer;
+  int rc = 0;
+  int reqtype = 0x00000005; //fullword. request type
+  int rectype = 0x0000004F; //SMF record type, only type 79 is supported
+  int subtype = 0x00000009; //SMF record subtype
+  int bufferlen = 716800; // Yes the SMF buffer is actually 700Kb roughly
+  int cpuUtil = 0x00000000;
+  int demandPaging = 0x00000000;
+  int options = 0x00000000;
+  int mvsSrm = 0x00000000;
+  int zaapUtil = 0x00000000;
+  int ziipUtil = 0x00000000;
+  uname(&unameRet);
+  gethostname(hostnameBuffer, sizeof(hostnameBuffer));
+  buffer = (char *)safeMalloc(bufferlen, "buffer");
+  memset(buffer, 0, bufferlen);
+  smfFunc = (EXSMFI *)fetch("ERBSMFI");
+  rc = (*smfFunc)(&reqtype,
+               &rectype,
+               &subtype,
+               buffer,
+               &bufferlen,
+               &cpuUtil,
+               &demandPaging,
+               &options,
+               &mvsSrm,
+               &zaapUtil,
+               &ziipUtil);
+  if(rc > 0){
+    respondWithError(response, HTTP_STATUS_BAD_REQUEST, "Unable to fetch from RMF data interface service");
```

Comment:
and then free the buffer and return, right?

Reply:
I free the buffer above line 156 now, its never used after and does not contain (human) readable data.  respondWithError returns the response.

Reply:
but we still need to return from the function now